### PR TITLE
MBS-11882 Fix ToastRule in Android 11 target

### PR DESCRIPTION
**Pull request by eugene-krivobokov**
Breaking change in toasts for apps with targetSdk 30 in Android 11:
No view for toasts created by `Toast.makeText`.

https://developer.android.com/about/versions/11/behavior-changes-11#toasts

### CODE REVIEW

File: subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ProxyToastChecks.kt [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import org.hamcrest.Matchers.empty
+import org.hamcrest.Matchers.not
+import org.hamcrest.TypeSafeMatcher
+
+internal class ProxyToastChecks(
+    private val proxyToast: MockProxyToast,
+    private val textAccessor: ToastTextAccessor,
+) : ToastChecks {
+
+    private val longToastDuration = 3500L
```
Comment (dsvoronin) [inline subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ProxyToastChecks.kt]: Any chance to get value from android resources?

File: subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ProxyToastChecks.kt [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+    override fun toastDisplayedWithText(text: String) {
+        internalToastDisplayedWithText(text)
+    }
+
+    override fun toastDisplayedWithText(@StringRes resId: Int) {
+        val toast = firstToastOrThrow()
+        val context = toast.getFieldByReflection("mContext") as Context
```
Comment (sergioboi) [inline subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ProxyToastChecks.kt]: Are you sure we should use reflection here?
Could we simply use context from `Instrumentation`?

File: subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ProxyToastChecks.kt [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+            MatcherAssert.assertThat(
+                "There are no toasts shown",
+                proxyToast.shownToasts,
+                not(empty())
+            )
+            proxyToast.shownToasts.first()
+        }
+    }
+
+    private fun withView(viewMatcher: Matcher<View>): Matcher<Toast> = object : TypeSafeMatcher<Toast>() {
```
Comment (sergioboi) [inline subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ProxyToastChecks.kt]: Why do you think that we don't need a class?

File: subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ToastTextAccessor.kt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+internal class InAppToastTextAccessor : ToastTextAccessor {
+
+    override fun text(toast: Toast): String? {
+        @Suppress("DEPRECATION")
+        val view = requireNotNull(toast.view) {
+            "Toast.view must be present in $toast"
+        }
+        val messageView: TextView? = view.findViewById(android.R.id.message)
+        val isCustomView = messageView == null
+        return if (isCustomView) null else messageView!!.text.toString()
```
Comment (sergioboi) [inline subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ToastTextAccessor.kt]: force unwrap should be skipped

File: subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ToastTextAccessor.kt [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    }
+}
+
+internal class SystemUIToastTextAccessor : ToastTextAccessor {
```
Comment (sergioboi) [inline subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ToastTextAccessor.kt:60]: I need a comment about how Toast draws in 30 and upper APIs

File: subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ToastTextAccessor.kt [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+    override fun text(toast: Toast): String? {
```
Comment (sergioboi) [inline subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ToastTextAccessor.kt:62]: Let's discuss a behavior.  We are trying to express the complex state via nullabillity

**Review by sergioboi — APPROVED** [review]

File: subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ProxyToastChecks.kt (hunk 1 above)
Comment (eugene-krivobokov) [reply subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ProxyToastChecks.kt]: Found it in a hidden class. 
We could access it only through reflection, but it will be version-specific.
I think it's too much for this knowledge.

File: subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ProxyToastChecks.kt (hunk 2 above)
Comment (eugene-krivobokov) [reply subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ProxyToastChecks.kt]: Thanks, didn't think of it. 
Fixed

To be correct, application context is not a complete replacement. There is a case of configuration-specific resources. 
But it looks like a corner case. We can live without it.

File: subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ProxyToastChecks.kt (hunk 3 above)
Comment (eugene-krivobokov) [reply subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ProxyToastChecks.kt]: Fixed, introduced a class

File: subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ToastTextAccessor.kt (hunk 4 above)
Comment (eugene-krivobokov) [reply subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ToastTextAccessor.kt]: Fixed

File: subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ToastTextAccessor.kt (hunk 5 above)
Comment (eugene-krivobokov) [reply subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ToastTextAccessor.kt:60]: Added

File: subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ToastTextAccessor.kt (hunk 6 above)
Comment (eugene-krivobokov) [reply subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ToastTextAccessor.kt:62]: Added an explicit check for the existence of the field.
Besides that, it looks ok. The field is nullable in the case of a custom view.
